net: socket: packet: Add EtherCAT protocol support

Zhiqiang-Hou opened the pull request:
Add EtherCAT protocol support, now applications can
transmit/receive EtherCAT packets via RAW socket.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou committed:
net: socket: packet: Add EtherCAT protocol support

Add EtherCAT protocol support, now applications can
transmit/receive EtherCAT packets via RAW socket.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

jukkar commented on subsys/net/lib/sockets/sockets_packet.c:
```diff
@@ -386,7 +386,8 @@ static const struct socket_op_vtable packet_sock_fd_op_vtable = {
 {
 	if (((type == SOCK_RAW) && (proto == ETH_P_ALL)) ||
-		((type == SOCK_RAW) && (proto == IPPROTO_RAW)) ||
+	    ((type == SOCK_RAW) && (proto == IPPROTO_RAW)) ||
+	    ((type == SOCK_RAW) && (proto == ETH_P_ECAT)) ||
```
This change should be introduced after we have something that can receive EtherCAT frames. Do you have plans to submit the actual EtherCAT stack implementation?

Zhiqiang-Hou replied:
The EtherCAT frames are filled in application layer, then use socket (domain: PF_PACKET and type: SOCK_RAW) to pass them to ethernet driver to send out without changing; the reverse process in receive direction. So nothing is needed in Zephyr networking stack but check the protocol, and this patch is just to let the socket interface permit this kind of packets pass through.

No plan, currently I'm using SOEM, but wondering if it can be integrated into Zephyr.

Thanks,
Zhiqiang

jukkar replied:
Ok, makes sense.

jukkar commented on include/net/ethernet.h:
```diff
@@ -65,6 +65,7 @@ struct net_eth_addr {
 #define NET_ETH_PTYPE_PTP		0x88f7
 #define NET_ETH_PTYPE_LLDP		0x88cc
 #define NET_ETH_PTYPE_ALL               0x0003 /* from linux/if_ether.h */
+#define NET_ETH_PTYPE_ECAT		0x88a4
```
Please call this `NET_ETH_PTYPE_ETHERCAT`

Zhiqiang-Hou replied:
SOME uses the pattern 'ECAT', I think it's brief and not ambiguous. it must be changed to 'ETHERCAT'?

jukkar replied:
Hmm, I see both being used in different opensource implementations. Lets go with shorter one then as that is used in SOEM.

jukkar commented on include/net/ethernet.h:
```diff
@@ -84,6 +85,9 @@ struct net_eth_addr {
 #define ETH_P_TSN	NET_ETH_PTYPE_TSN
 #endif
+#if !defined(ETH_P_ECAT)
+#define  ETH_P_ECAT	NET_ETH_PTYPE_ECAT
```
And here `ETH_P_ETHERCAT`

Zhiqiang-Hou replied:
Ditto

Zhiqiang-Hou commented:
Hello @rlubos @jukkar,

Can you help me understanding how this patch results in these 2 twister-build failures?
"Run tests with twister / twister-build (6) (pull_request_target) Failing after 35s — twister-build (6)"
"Run tests with twister / twister-build (10) (pull_request_target) Failing after 35s — twister-build (10)"

Thanks,
Zhiqiang

rlubos commented:
@Zhiqiang-Hou The previous failure was due to environmental issue (west failed to fetch one of the repositories). I tried to rerun, but now it fails with:
```
Your branch is not up to date, you need to rebase on top of latest HEAD of main branch
```
Please, rebase your branch.

Zhiqiang-Hou commented:
> @Zhiqiang-Hou The previous failure was due to environmental issue (west failed to fetch one of the repositories). I tried to rerun, but now it fails with:
> 
> ```
> Your branch is not up to date, you need to rebase on top of latest HEAD of main branch
> ```
> 
> Please, rebase your branch.

Rebased. Thanks a lot!

> @Zhiqiang-Hou The previous failure was due to environmental issue (west failed to fetch one of the repositories). I tried to rerun, but now it fails with:
> 
> ```
> Your branch is not up to date, you need to rebase on top of latest HEAD of main branch
> ```
> 
> Please, rebase your branch.

Rebased. Thanks a lot!
